### PR TITLE
Add dilation parameter and tests

### PR DESCRIPTION
**Pull request by aretor**
The implementation now allows for the setting of a dilation parameter. The tests have been run without any problem.

### CODE REVIEW

**Review by fkodom — APPROVED** [review]
Excellent!  Thanks so much for this contribution -- I'll get this merged over and tag a new release 🎉

**Comment by aretor** [comment]
You welcome :)

**Comment by yoyolicoris** [comment]
@aretor @fkodom Sorry for bothering you guys, but I think we should roll back this merge or make a quick fix.
The zero-insertion should be placed at the `forward` call, not in the `__init__`, otherwise those inserted zeros will be optimized jointly.

**Comment by aretor** [comment]
Yes, that's right. What about adding a backward hook to `self.weight` in the `__init__` that takes care of zeroing unwanted gradients?

**Comment by yoyolicoris** [comment]
> Yes, that's right. What about adding a backward hook to `self.weight` in the `__init__` that takes care of zeroing unwanted gradients?

Hmm, this is one way. But I still think zeros should not be included as parameters at the very beginning.

I would recommend doing something like this:

in `__init__`
```python
if any(d != 1 for d in dilation):
    total_size = tuple(
        ((ks - 1) * dil + 1)
        for ks, dil in zip(kernel_size, dilation)
    )
    mask = torch.zeros(out_channels, in_channels // groups, *total_size, dtype=torch.bool)
    ids = tuple(
        torch.arange(0, tot_sz, dil)
        for tot_sz, dil in zip(total_size, dilation)
    )
    mask[(slice(None), slice(None),) + torch.meshgrid(*ids)] = 1
    self.register_buffer('dilation_mask', mask)
```

in `forward`

```python
if hasattr(self, 'dilation_mask'):
    weight = signal.new_zeros(self.dilation_mask.shape)
    weight.masked_scatter_(self.dilation_mask, self.weight)
else:
    weight = self.weight
# resume to call fft_conv but replace `self.weight` with `weight`
```

BTW, the above operations are quite fit in the use of [parameterize](https://pytorch.org/docs/1.9.0/generated/torch.nn.utils.parametrize.register_parametrization.html#torch.nn.utils.parametrize.register_parametrization) module, but it will be a breaking change since it was introduced after torch 1.9.0.

**Comment by aretor** [comment]
The idea makes sense, maybe we can store the mask indexes `ids` for saving memory, though `masked_scatter_` should be more time-efficient.

The `torch.register_hook` idea is less verbose and consists of adding in the `__init__` the following line: `weight.register_hook(lambda grad: grad * mask)` replacing the gradient with its masked version.

IMHO I consider both the solutions to be temporary. In the long term, it could be more appropriate to write part of the module with C++ LibTorch.

Finally, it would be a good idea to add a unit test on the backward pass and a `weight` property that returns the non-zero elements, in order to make the behavior more consistent with PyTorch `Conv*d` modules.

**Comment by fkodom** [comment]
Sorry, I missed this conversation somehow.  Will look into these updates in the near future 👍

**Comment by aretor** [comment]
I experimented a bit in the past days and, following our considerations, I came to a possible solution, that replaces the dilation padding with a Kronecker product. I have submitted a [pull request](https://github.com/fkodom/fft-conv-pytorch/pull/13), which can replace the previous release.

I have added some new unit tests on the gradients to check the validity of the proposed solution. Perhaps, they can be expanded to check the approximate equivalence of a full optimization cycle between fft and normal convolution.
